### PR TITLE
Validate plugins removed from hub before marking them stale

### PR DESCRIPTION
**Pull request by manasaV3**
## Description
Relates to: https://github.com/chanzuckerberg/napari-hub/issues/1275

In cases where a plugin is being marked stale without a new version, we validate if the version is no longer active by validating it against `npe2api` repository's `public/classifiers.json`. 

### TODO:
Update the documentation to reflect it can now take upto 2 hours for a plugin to be removed from the site.

### CODE REVIEW

File: .happy/terraform/modules/ecs-stack/main.tf [hunk 1]
```diff
@@ -341,6 +341,7 @@ module data_workflows_lambda {
     "GITHUB_CLIENT_ID"      = local.github_client_id
     "GITHUB_CLIENT_SECRET"  = local.github_client_secret
     "PLUGINS_LAMBDA_NAME"   = local.plugins_function_name
+    "ZULIP_CREDENTIALS"     = local.zulip_credentials
```
Comment (DragaDoncila) [inline .happy/terraform/modules/ecs-stack/main.tf:344]: Is this re-enabling the zulip stream @manasaV3? Perhaps it would be good to do that in a separate PR once we've merged this and confirmed all is working as intended? Ideally tbh I would like to actually capture one of the events reported in #1725 before we re-enable.

Also regardless of that, we should flag to folks on zulip before we re-enable the channel.

File: data-workflows/plugin/tests/test_classifier_adapter.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        self._github_client_helper.get_file.assert_called_once_with(FILE_NAME, "json")
+
+    def test_handle_invalid_query_data(self):
+        self._classifier_json = {"inactive": []}
+        assert False == plugin.classifier_adapter.is_plugin_live("foo", "1.0.2")
+        assert False == plugin.classifier_adapter.is_plugin_live("bar", "1.0.2")
+        self._github_client_helper_call.assert_called_once_with(REPOSITORY)
+        self._github_client_helper.get_file.assert_called_once_with(FILE_NAME, "json")
+
+    def test_is_plugin_live_caching(self):
```
Comment (DragaDoncila) [inline data-workflows/plugin/tests/test_classifier_adapter.py:49]: I don't really understand how this is different from `test_handle_valid_query_data`

File: data-workflows/plugin/processor.py [hunk 3]
```diff
@@ -39,15 +40,19 @@ def _is_new_plugin(plugin_version_pair):
...
-                plugin_metadata_type=PluginMetadataType.PYPI,
-            )
-            if pypi_plugin_version is None:
-                zulip.plugin_no_longer_on_hub(name)
+        if pypi_plugin_version == version or (
+            pypi_plugin_version is None and is_plugin_live(name, version)
+        ):
+            continue
+
+        logger.info(f"Updating old plugin={name} version={version}")
```
Comment (DragaDoncila) [inline data-workflows/plugin/processor.py:48]: What other events (besides a plugin being removed) reach this line of code? The comment says `"update for removed plugins and existing older version of plugins"`, but afaik we only update metadata when a plugin is first released?

File: data-workflows/plugin/classifier_adapter.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    return data
+
+
+def is_plugin_live(name: str, version: str) -> bool:
+    try:
+        recent_query_update = _get_recent_query_data()
+        active_versions = set(recent_query_update.get("active", {}).get(name, []))
+        return version in active_versions
+    except RuntimeError:
+        LOGGER.warning(f"Returning {name} {version} is live due to RuntimeError")
```
Comment (DragaDoncila) [inline data-workflows/plugin/classifier_adapter.py:27]: would be good to actually print out the message from the `RuntimeError` here

File: data-workflows/plugin/tests/test_classifier_adapter.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        self._classifier_json = {"active": {"foo": ["1.0.0", "1.0.1", "1.0.2"]}}
+        assert True == plugin.classifier_adapter.is_plugin_live("foo", "1.0.2")
+        assert False == plugin.classifier_adapter.is_plugin_live("foo", "1.0.4")
+        self._github_client_helper_call.assert_called_once_with(REPOSITORY)
+        self._github_client_helper.get_file.assert_called_once_with(FILE_NAME, "json")
+
+    def test_is_plugin_live_does_not_cache_error(self):
+        self._classifier_json = None
```
Comment (DragaDoncila) [inline data-workflows/plugin/tests/test_classifier_adapter.py:57]: would be good to add a comment here, this test is very opaque.

File: .github/workflows/backend-tests.yml [hunk 6]
```diff
@@ -47,6 +47,7 @@ jobs:
       - name: Install dependencies
         # need npe2 here to test preview page
         run: |
+          sudo apt install libffi7
```
Comment (DragaDoncila) [inline .github/workflows/backend-tests.yml:50]: What prompted this change?

**Review by DragaDoncila — COMMENTED** [review]
@manasaV3 thanks for working on this! Left a couple of comments.

File: .github/workflows/backend-tests.yml (hunk 6 above)
Comment (manasaV3) [reply .github/workflows/backend-tests.yml:50]: This was to fix issues with ubuntu's latest distro not having the required files, which was causing the tests to fail. I should move this patch to a separate PR.

File: data-workflows/plugin/classifier_adapter.py (hunk 4 above)
Comment (manasaV3) [reply data-workflows/plugin/classifier_adapter.py:27]: As the run time error generated [here](https://github.com/chanzuckerberg/napari-hub/pull/1301/files#diff-d3d31a1bbe0f387eac4ae3570367d4a7ae2c112ce5d4dce794883d1d5c376bd5) doesn't provide any additional information, I don't think it would be useful to log it. 

Instead, updating the `github_client_helper.get_file(..)` to print the exception before it handles it gracefully will provide the context.

File: data-workflows/plugin/tests/test_classifier_adapter.py (hunk 2 above)
Comment (manasaV3) [reply data-workflows/plugin/tests/test_classifier_adapter.py:49]: You are right, this is a duplicate test that should be removed.

File: .happy/terraform/modules/ecs-stack/main.tf (hunk 1 above)
Comment (manasaV3) [reply .happy/terraform/modules/ecs-stack/main.tf]: Agreed that we should decouple adding the Zulip credentials from this PR. Opened [this PR](https://github.com/chanzuckerberg/napari-hub/pull/1304) for that. 

We could let this run in prod for up to a month to see if we can catch the issue occurring. But, if we aren't able to, I think we could add Zulip notifications to let the community know that we are cautiously positive about having fixed the issue. What do you think?

File: data-workflows/plugin/classifier_adapter.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+FILE_NAME = "public/classifiers.json"
+LOGGER = logging.getLogger(__name__)
+
+
+@cache
+def _get_recent_query_data() -> dict:
+    github_client_helper = GithubClientHelper(REPOSITORY)
+    data = github_client_helper.get_file(FILE_NAME, "json")
+    if not data:
+        raise RuntimeError(f"Unable to fetch {FILE_NAME} from {REPOSITORY}")
```
Comment (DragaDoncila) [inline data-workflows/plugin/classifier_adapter.py:17]: might be worth logging here?

File: data-workflows/plugin/processor.py (hunk 3 above)
Comment (DragaDoncila) [reply data-workflows/plugin/processor.py:52]: Ah yeah of course. Thanks for clarifying!

File: .happy/terraform/modules/ecs-stack/main.tf (hunk 1 above)
Comment (DragaDoncila) [reply .happy/terraform/modules/ecs-stack/main.tf]: @manasaV3 yep I think that sounds reasonable!

**Review by DragaDoncila — APPROVED** [review]
@manasaV3 looks good to me - just left a small comment about perhaps logging when the `classifiers` file is unavailable
